### ISSUE TITLE
Index as we mux

### ISSUE THREAD

**Issue by davedoesdev** [davedoesdev/webm-muxer.js#15]
Can EBML.js index as it goes and not at the end?
This would require the index to go at the end of the WebM file.

**Comment by sashazxtt** [comment]
I vote for this (if such a thing as voting exists 😁)

**Comment by davedoesdev** [comment]
This would need `SeekHead` and `Cues` to be written at the end, which isn't recommended but it's not a SHOULD or MUST not in the spec. Needs looking at EBML.js to see if it can be persuaded to do that (driven by an option ideally).

**Comment by davedoesdev** [comment]
I looked at libwebm mkvmuxer, if the writer isn't seekable then it doesn't output any seek info (which is why EBML.js to be used to fix it up).

**Comment by davedoesdev** [comment]
Interestingly, just putting the `refinedMetadata` at the end still seems to generate a valid WebM file

```js
new Blob([body, refinedMetadataBuf]
```

I don't have time to test it more right now - it will require removing the buffering of all the output and checking it still plays and seeks.

**Comment by davedoesdev** [comment]
I'm not sure this is going to be possible: https://www.matroska.org/technical/ordering.html

> The first Info Element and the first Tracks Element MUST either be stored before the first Cluster Element or both SHALL be referenced by a SeekHead Element occurring before the first Cluster Element

The info element contains the duration so we'd not be able to fix that up.

> If used, the first SeekHead Element SHOULD be the first non-CRC-32 Child Element of the Segment Element. If a second SeekHead Element is used, then the first SeekHead Element MUST reference the identity and position of the second SeekHead

So the SeekHead (which references the other top-level elements) should be the first element.
We'd have to try omitting SeekHead.

**Comment by davedoesdev** [comment]
Just the info element having to be placed at the start is enough to stop this working - there would be no duration on the file.

**Comment by davedoesdev** [comment]
I tried putting the metadata at the end but the players said the file was empty. I used mkvinfo -a to check the parts were there.

**Comment by davedoesdev** [comment]
Closing as wontfix - I don't think it's possible to have a working webm file with the info and seekhead elements at the end.

**Comment by davedoesdev** [comment]
Just a note to anyone coming after: If a duration in the file isn't required, it might be possible to put just the SeekHead and Cues at the end. See makeMetadataSeekable in EBML.js

**Comment by sashazxtt** [comment]
Thanks for the investigations, Dave.
There is a webm js writer that I used that does produce seekable files. They do it by seeking to some positions close to the beginning of the file and writing the cues there. The code for it is accessible from the Webcodecs github project (IIRC; let me see if I can dig it up).

**Comment by sashazxtt** [comment]
Here is the webm-writer source. I think the files produced with it (video only, no audio) did have some issues with playback, but they do have duration and are seekable.
https://w3c.github.io/webcodecs/samples/capture-to-file/webm-writer2.js

**Comment by davedoesdev** [comment]
Yes rewriteDuration() seeks to the start, using fs (require('fs')). But can
we do that in the browser?

**Comment by davedoesdev** [comment]
Maybe something like
https://developer.mozilla.org/en-US/docs/Web/API/FileSystemWritableFileStream
?

**Comment by sashazxtt** [comment]
Ah, I'm using Electron, so have full file access.
Yes, using FileSystemWritableFileStream will work too. In my experiments I've wrapped node's fs into that interface and it worked well.

**Comment by davedoesdev** [comment]
OK so I think this might be doable, I'll leave this issue open and take a look at some point

**Comment by stanislavys** [comment]
One implementation option to consider is to use libwebm's Segment `kFile` mode. The underlying `MkvMuxer` seems to implement the behavior above (overwrite the duration in the SegmentInfo element and append cues at the end of the segment) in that mode. Currently the `WebMLiveMuxer` facade uses `kLive` mode. So perhaps webm-muxer's API could allow the client to select the desired mode for the output, assuming the supplied persistence sink has sufficient capabilities (i.e. seekable), and webm-muxer could pick the appropriate libwebm "wrapper" accordingly.

**Comment by stanislavys** [comment]
To follow the WebM guideline that the Cues should be before any clusters, there is the `Segment::CopyAndMoveCuesBeforeClusters` helper which could be used after the Segment has been finalized and as an additional optimization step.

**Comment by davedoesdev** [comment]
Thanks, that's interesting. I'd have to implement an `IMkvWriter` which sent a message back to JS to seek in the stream.
It may be easier to follow a similar approach to the existing code using EBML but using `FileSystemWritableFileStream` instead of in-memory buffers. At least I'll try that first.

**Comment by stanislavys** [comment]
So basically switch to `FileSystemWritableStream` for the persistence and have a post-processing step which uses EBML.js to get the cues and use `FileSystemWritableStream` to do the right patches?

**Comment by sashazxtt** [comment]
Yes, that will work - it did when  I used FileSystemWritableStream wrapper around the node fs API with webm-writer.

**Comment by davedoesdev** [comment]
Yes that's the idea. Btw, thanks for both your help, appreciate it.

**Comment by davedoesdev** [comment]
https://github.com/davedoesdev/webm-muxer.js/commit/81bc8c3487a93f1589a040db043d8659fb70f7a4

What do you think? Due to having to put Cues at the start and because EBML uses varints, it's not possible just to seek to the start and overwrite the metadata. It seems `CopyAndMoveCuesBeforeClusters` does similar.

I guess one option would be to reserve some space before the first cluster and overwrite it if the Cues and revised header fit?

**Comment by stanislavys** [comment]
Yes, that looks like a good pragmatic solution. The one downside I see is that it might not scale well with larger recordings - e.g. extra parse step for the muxer output, duplicate the required space on disk for the original and the patched output, not sure of EBML's performance for larger files too. These should not be a problem in most cases and can be further optimized if necessary.

> Due to having to put Cues at the start and because EBML uses varints, it's not possible just to seek to the start and overwrite the metadata.

If Cues are appended at the end, seeking and overwriting the meta data at the beginning should be possible if sufficiently large VINTs are used (e.g. >=4 bytes). Same for the duration which IIRC is a float. I think the EBML schema allows variance in the binary size of these metadata fields.

**Comment by davedoesdev** [comment]
I'm considering writing empty space at the start before muxing - enough for the metadata including SeekHead.
Then append the cues.
Then fix up the metadata in-place to point to where the cues were appended.

**Comment by stanislavys** [comment]
Sounds good. I'd be interested to see the outcome. I had done a back of napkin calculation for the necessary placeholder size for such an approach and it seems like a few hundred kb should be enough for recordings up to 8 hours with video keyframes every 5s. There is probably a function of expected max duration and keyframe rate (assuming video is always present in the output) which can be used to come up with a comfortable estimate such that element relocation is not required.

**Comment by stanislavys** [comment]
How do you plan to handle any remaining space in the placeholder after writing the actual cues, if any? Fill it with a Void element (i.e. ensure it is accounted for in the estimated size and the allowed cue size to write)?

**Comment by davedoesdev** [comment]
@stanislavys if I put the Cues at the end, would the placeholder size depend on the size of the recording any more?

**Comment by davedoesdev** [comment]
Regarding the remaining space - do players attempt to parse it if I put in a SeekHead or would they just read from the positions in the SeekHead?

**Comment by davedoesdev** [comment]
Is the Void element a single byte (0xEC)?

**Comment by stanislavys** [comment]
> if I put the Cues at the end, would the placeholder size depend on the size of the recording any more?

I would not expect the placeholder size to vary in such case. It's a matter of reserving big enough elements in the SeekHead for the Cues position and in the SegmentInfo for the duration.

> Regarding the remaining space - do players attempt to parse it if I put in a SeekHead or would they just read from the positions in the SeekHead?

I am not sure. I would expect the content of the element is ignored. At least the EBML spec description seems to imply that: https://datatracker.ietf.org/doc/html/rfc8794#section-11.3.2

> Is the Void element a single byte (0xEC)

I think it is an element with actual placeholder data whose length is reflected by the element.

**Comment by stanislavys** [comment]
BTW, the MkvMuxer seems to use a VoidElement which is eventually overwritten with the SeekHead. So maybe there is more than just offset-number-offsets...

```
// Reserves space by writing out a Void element which will be updated with
  // a SeekHead element later. Returns true on success.
  bool Write(IMkvWriter* writer);
```

Implementation:
```
bool SeekHead::Write(IMkvWriter* writer) {
  const uint64_t entry_size = kSeekEntryCount * MaxEntrySize();
  const uint64_t size =
      EbmlMasterElementSize(libwebm::kMkvSeekHead, entry_size);

  start_pos_ = writer->Position();

  const uint64_t bytes_written = WriteVoidElement(writer, size + entry_size);
  if (!bytes_written)
    return false;

  return true;
}
```

**Comment by davedoesdev** [comment]
I guess the metadata size would only vary very slightly depending on how big the VINT is in the SeekHead which points to the Cues at the end. But that would only be a few bytes.

I'll somehow need to arrange it such that there is an exact multiple of Void-sized bytes left.

**Comment by davedoesdev** [comment]
Actually I realise from that code that I don't need to have extra remaining space.
The size of the SeekHead is predictable (as in `SeekHead::Write`).

**Comment by davedoesdev** [comment]
The `SeekHead::Finalize()` code ends up writing the actual data using `GetIntSize()` so surely the written size of the SeekHead is smaller than that which was reserved. In which case the left over space will be garbage - maybe players just ignore that.

**Comment by davedoesdev** [comment]
This is interesting to read: https://sourceforge.net/p/mediainfo/bugs/360/

> > Unfortunatly, in this file there is a NULL byte before the next EBML
> > element after the first "SeekHead", so MediaInfo stops (it does not
> > understand this NULL byte).
> > When I remove this NULL byte (in your file, offset 0x66), All is OK
> 
> Such NULL bytes are actually OK. They can happen due to the following
> reason:
> 
> If an application tries to overwrite an existing element in a Matroska
> file with new content and the content is smaller than it was before then
> there's "empty" space left behind the element. If that empty space is
> five bytes long or longer than the application can write an EbmlVoid
> element into this gap. However, if the space is only four bytes or less
> then the EbmlVoid element does not fit. Therefore the space must be
> overwritten with NULL bytes.
> 
> Your application should skip NULL bytes when finding the next element's
> ID if the NULL byte is the first byte of the ID.

**Comment by davedoesdev** [comment]
But then:

> I don't see this anywhere in the specs.
> Trusting http://www.matroska.org/technical/specs/index.html , a NULL byte is not defined (1xxxxxxx to 00000001 are defined, -1 is defined too, but not 00000000!)
> I am not sure this is wanted to have NULL byte saying "skip me".
> Specifications say: "larger Element Sizes are left undefined at the moment", and this specification can say later, to be logic, that 00000000 1xxxxxxx 2-byte code if for >2^56 sizes!
> 
> I understand the problem of padding, but from my point of view the SeekHead element size should include this padding byte (ie: in your file, SeekHead element size should be 0xBE instead of 0xBD, so the padding is inside the SeekHead element, at the end so no need of skiping NULL byte, just discard the end, and the next byte in the Segment element is the next sub-level of the segment), as .mp4 does with atoms.

**Comment by davedoesdev** [comment]
```c++
    const uint64_t total_entry_size = kSeekEntryCount * MaxEntrySize();
    const uint64_t total_size =
        EbmlMasterElementSize(libwebm::kMkvSeekHead, total_entry_size) +
        total_entry_size;
    const int64_t size_left = total_size - (writer->Position() - start_pos_);

    const uint64_t bytes_written = WriteVoidElement(writer, size_left);
    if (!bytes_written)
      return false;
```

It fills the remaining space with a Void.

**Comment by davedoesdev** [comment]
So my next question is what if `size_left` isn't big enough to hold a complete Void element?

**Comment by davedoesdev** [comment]
Surely `void_entry_size` will overflow?

```c++
uint64 WriteVoidElement(IMkvWriter* writer, uint64 size) {
  if (!writer)
    return false;

  // Subtract one for the void ID and the coded size.
  uint64 void_entry_size = size - 1 - GetCodedUIntSize(size - 1);
  uint64 void_size = EbmlMasterElementSize(libwebm::kMkvVoid, void_entry_size) +
                     void_entry_size;

  if (void_size != size)
    return 0;
```

**Comment by davedoesdev** [comment]
I guess that's not my problem. I can figure out the max size of the SeekHead and hence the metadata, and make sure there's plenty left over to insert the void after it.

**Comment by stanislavys** [comment]
Exactly. This is what I was referring to in my earlier comment with "ensure it is accounted for in the estimated size". I was thinking when reserving space, you could also reserve the minimum size VoidElement (would that be 3 bytes - one for ID, 1 for len and one for a data element (perhaps the last one can be skipped)). That way, the placeholder will always end with a valid VoidElement which is spec compliant and should be ignored by players.

**Comment by davedoesdev** [comment]
OK I follow what you mean thanks.
Won't the fixed up SeekHead differ in size depending on the VINT positions in it? So like `SeekHead::Finalize`, we'll need to put in a Void element after it - we won't be able to statically allocate one in the placeholder.

**Comment by stanislavys** [comment]
Right, the remaining space would vary and will have to be filled with a VoidElement of the right size. As long as a minimum size VoidElement can fit, it can be expanded to cover the entire remaining space.

**Comment by davedoesdev** [comment]
OK I think I have a plan, I'll see if I can get some time to implement it in the next few days. Thanks again.

**Comment by davedoesdev** [comment]
OK I implemented the plan and the code is now on the main branch.
See https://github.com/davedoesdev/webm-muxer.js/blob/main/webm-writer.js
I'll add some docs for it when I get time.

**Comment by davedoesdev** [comment]
It'll try to put the cues before the clusters if there's space, otherwise it puts them at the end.
It rewrites the header only (by seeking to the start of the file) so the data isn't copied again.
The amount of reserved space is configurable (defaults to 1024 bytes).
I had to make some modifications to `EBML.js` which should probably be made in the original TypeScript source instead, and possibly upstreamed (though it's a different use case).

**Comment by davedoesdev** [comment]
I opted to remove the version which buffers all the data in memory but perhaps it might be useful to someone.
Feedback appreciated.

**Comment by sashazxtt** [comment]
Fantastic work, Dave! I will give that a try!

**Comment by davedoesdev** [comment]
Great thanks. I think I might put the in-memory code back in but an as
optional alternative, keeping the file system version as the default.

**Comment by stanislavys** [comment]
Thanks for the quick turnaround on this. I reviewed the code and I think it looks great. I liked the ability to specify the size of the placeholder. Keeping both persistence options available would be really nice.
BTW, when I tried the demo at https://rawgit-now.netlify.app/davedoesdev/webm-muxer.js/main/demo.html, the stop button remained inactive and I wasn't able to stop the recording. Is this a known issue?

**Comment by davedoesdev** [comment]
I tried the demo and it worked for me. Are there any errors in the console?

**Comment by davedoesdev** [comment]
OK I put an option in to write into memory like before (just don't pass in a suggested filename to the writer).
@stanislavys  Once we get to the bottom of your issue with the stop button, I'll update the docs and close this issue.

**Comment by davedoesdev** [comment]
Documentation updated to include webm-writer.js

**Comment by davedoesdev** [comment]
@stanislavys occasionally the tests fail with the same symptom you're seeing on the demo (timeout trying to stop). However, no logs to say why. If you're still seeing the issue, it'd be great to trying to get to the bottom of it.

**Comment by stanislavys** [comment]
Apologies for the delay, @davedoesdev. I just reran the demo and it appears the problem is on my side. I used Chrome 93 without toggling the switch to enable experimental features. Here is the console output:
```
demo.js:94 Uncaught (in promise) ReferenceError: MediaStreamTrackProcessor is not defined
    at HTMLInputElement.<anonymous> (demo.js:94)
(anonymous) @ demo.js:94
async function (async)
(anonymous) @ demo.js:81
```

Testing in Chrome Canary worked as expected. It seemed like in Chrome 93 the recording did not start at all but the start button got deactivated regardless and also the camera got activated (the camera light was on all the time the site was open), giving the impression that the recording is in progress. Perhaps, the UI should be reverted back to its original state, or not update at all, in case of such errors. Although, with the official release of Insertable Streams, that might not be a common problem anymore.
On the topic of stop timing out, we've been looking into an issue where the read() operation on a reader obtained from `MediaStreamProcessor.readable` does not resolve when the stream it reads from is ended. One scenario to reproduce this is to obtain a screen capture stream for an application (as opposed to the entire screen) and quit the app. In my understanding of the docs, the read operation is supposed to resolve with a result with `done` set to `true` but that does not seem to happen. Could the issue you are observing be related to this? Any ideas what might be causing this in general?
For such ended streams, trying to stop the encoder workers seem to fail because of the non-resolving promise of the `read()` operation.

**Comment by davedoesdev** [comment]
@stanislavys thanks for the update. As it's just a demo I won't bother catching exceptions for now. I updated the docs to say it requires Chrome 95.

If you're seeing what you describe then that would cause the issue in the test. Sounds like a bug in Chrome.

Thanks again for all your help.

**Comment by sashazxtt** [comment]
Finally, I was able to pick up the changes.
I had to massage the WebmWriter a bit to fit our use case (just straight write, no buffering options).
I get files that are seekable. But... they are playable by VLC only. Some files (I think shorter in duration) are playable by the Windows player. Another issue is that the playback doesn't start right away, but with some delay. Without writing the cues section the playback starts properly.
Maybe the issue is in my massaging, but so far I don't see where.

**Comment by sashazxtt** [comment]
OK, it appears it was a bit of a cock-up on my side - in removing the save file picker dialog I removed the code that reserves the space at the start of the file.
Here is what I see during the playback.
 - VLC shows 6 streams in the file - three of each audio and video
 - VLC seeking seems to be granular at the key frame interval
 - VLC doesn't display the duration
 - Chrome plays the file OK and seeking is granular to the point of the mouse click on the seekbar, but only after it had a chance to parse the whole file. Eventually, the duration information is displayed as well.

**Comment by davedoesdev** [comment]
Reopening until I find time to repro

**Comment by davedoesdev** [comment]
@sashazxtt using the demo I can't repro any of those issues. VLC for me plays the recording fine. It shows 2 tracks and displays the duration.

**Comment by davedoesdev** [comment]
Chrome is also fine - shows duration and seeks to mouse click straight away

**Comment by sashazxtt** [comment]
Here is ffprobe tells me:
[matroska,webm @ 000002ac9dcb4c40] 0x00 at pos 272 (0x110) invalid as first byte of an EBML number
[matroska,webm @ 000002ac9dcb4c40] Duplicate element
[matroska,webm @ 000002ac9dcb4c40] Element at 0x41 ending at 0x114e00 exceeds containing master element ending at 0x13ed
[matroska,webm @ 000002ac9dcb4c40] Duplicate element
[matroska,webm @ 000002ac9dcb4c40] Element at 0x4f ending at 0x2a9f exceeds containing master element ending at 0x13fb
[matroska,webm @ 000002ac9dcb4c40] Length 7 indicated by an EBML number's first byte 0x02 at pos 206 (0xce) exceeds max length 4.
[matroska,webm @ 000002ac9dcb4c40] Duplicate element
    Last message repeated 1 times
Input #0, matroska,webm, from '1.webm':
  Metadata:
    encoder         : libwebm-0.3.0.0
  Duration: N/A, start: 0.000000, bitrate: N/A
  Stream #0:0(eng): Video: vp8, yuv420p(progressive), 1920x1080, SAR 1:1 DAR 16:9, 15 fps, 15 tbr, 1k tbn, 1k tbc (default)
  Stream #0:1(eng): Audio: opus, 48000 Hz, mono, fltp (default)
  Stream #0:2(eng): Video: vp8, yuv420p, 1920x1080, SAR 1:1 DAR 16:9, 15 fps, 15 tbr, 1k tbn, 1k tbc (default)
  Stream #0:3(eng): Audio: opus, 48000 Hz, mono, fltp (default)
  Stream #0:4(eng): Video: vp8, yuv420p, 1920x1080, SAR 1:1 DAR 16:9, 15 fps, 15 tbr, 1k tbn, 1k tbc (default)
  Stream #0:5(eng): Audio: opus, 48000 Hz, mono, fltp (default)

I will keep at it to see if it's something i broke...

**Comment by sashazxtt** [comment]
With the demo app I produced a file that doesn't quite work in VLC and Windows player. It gets stuck on occasion (maybe triggered by seeking attempts).
In Chrome the same file seems to be working.

**Comment by sashazxtt** [comment]
[camera.zip](https://github.com/davedoesdev/webm-muxer.js/files/7263408/camera.zip)

**Comment by sashazxtt** [comment]
I think this code has a problem:
`       const space = this.options.metadata_reserve_size + this.reader.metadataSize;
        const has_space = () => {
            return refinedMetadataBuf.byteLength === space || refinedMetadataBuf.byteLength <= space - 2; // min Void size is 2
        };
`
In at least one instance I see that the refinedMetaDataBuf is larger than the reserved space. 
The start of the muxed data will be overwritten by the new metadata.
Surprisingly, such files play fine and ffprobe doesn't complain.

Should it be something like this:
`const spaceRequired = refinedMedataBuf.byteLength + reader.metadataSize;
const has_space = () => {
  return spaceRequired < reservedSpace;
` ?
The image shows a case where the refined metadata will overwrite the start of the muxed data.
![image](https://user-images.githubusercontent.com/23205112/135566384-4e4f3518-0bc4-4b70-b711-ab82ad780541.png)

File before writing the refined metadata:
![image](https://user-images.githubusercontent.com/23205112/135566572-212d15ef-b976-452c-bf04-e81804204007.png)

File after writing the metadata. 
![image](https://user-images.githubusercontent.com/23205112/135566639-3e0ccc51-e463-476a-808d-647a31be0829.png)

**Comment by sashazxtt** [comment]
Oh, btw that multiple streams issue was caused by my bug - I didn't realize that the `fs.write` doesn't adjust the file pointer when provided a `position` value. So, the important bits were written at the end of the file instead of the reserved area.

**Comment by davedoesdev** [comment]
@sashazxtt thanks for the update. Looks like something's overwriting something for sure. I'll take a look in the next few days. Thanks again.

**Comment by sashazxtt** [comment]
I think I get it now why the space is computed this way - it accounts for the initial metadata that is already present in the file and that should be skipped over. So overwriting the original metadata is fine.

**Comment by sashazxtt** [comment]
It appears that VLC has problems seeking. It always wants to seek to a previous video key frame. Even files cleaned with `mkclean` don't seek smoothly. With 10 seconds between key frames, VLC is not too pleasant to use. Chrome and even Windows player seek fine. MPV has no such problems either. 
Usually VLC works like a champ, but not here....

**Comment by davedoesdev** [comment]
Yes I've noticed VLC isn't as good. I've found 1 second key frames keeps it happier.

Are you still having issues with corrupted metadata?

**Comment by sashazxtt** [comment]
I've tried 1 sec key frames and that indeed helps. Maybe we'll use that option - at the cost of increased file size.
I think the corrupted metadata issue is resolved, thanks. The overwrite I saw is ok, as I now understand what's happening.
I tried to see if it is possible to have a single call to the 'makeSeekable' that would cover both cases of cues in front and in the back, but looking at its code I don't think that's possible.
As our use case will always create long recordings, I've opted to always place cues at the end, saving on one 'makeSeekable' call.

Thanks!
